### PR TITLE
Update Documentation for `wire_names` and `_Global`

### PR DESCRIPTION
**Pull request by csookim**
All changes are related to variable names, docstrings, and documentation. No changes to the logic.

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

### CODE REVIEW

**Comment by csookim** [comment]
Most of the updates are changes to variable names and docstrings. No changes to the logic.

**Review by alecandido — COMMENTED** [review]
In general, the PR seems alright, and I just have a relevant note about a single point.

However, if I'm correct, this is still partially lacking the explanation for a question that we already received, i.e. the role of `wire_names` in the qubit selection.
With this, I mean that we should clarify the following points:
1. `wire_names` is absolutely not limiting the qubits used for placing or routing, if these non-trivial passes are included in the transpiler - thus, using `wire_names` as input for a non-trivial placer/routing is only possibly helpful to improve the transpiler solution, but it won't prevent the usage of any qubit in the platform
2. `wire_names` are instead exactly honored by the Qibolab backend, when used directly (no transpilation), leading to compilation errors if the connectivity is insufficient
3. padding should not be required at any stage, since the information that used to provide is now fully replaced by the usage of `wire_names` and the `connectivity`, to the point that we should even be able to drop the `Preprocessing` (though this should be confirmed)

File: doc/source/code-examples/advancedexamples.rst [hunk 1]
```diff
@@ -145,6 +145,55 @@ however the user may create the full state as follows:
...
+
+    # Set the physical qubit names
+    circuit.wire_names = ["C", "A", "B", "D"]
+
+    # This is equivalent to:
+    # circuit = Circuit(4)
+    # circuit.add(gates.H("C"))
+    # circuit.add(gates.CNOT("C", "A"))
+    # circuit.add(gates.CZ("C", "B"))
+    # circuit.add(gates.M("B", "D"))
```
[the remark this answers is not in the document]
Comment (csookim) [reply doc/source/code-examples/advancedexamples.rst]: [6321b3c](https://github.com/qiboteam/qibo/pull/1531/commits/6321b3cf0c74eedd918192b51fd8603bfed0a1c8)

**Comment by csookim** [comment]
> In general, the PR seems alright, and I just have a relevant note about a single point.
> 
> However, if I'm correct, this is still partially lacking the explanation for a question that we already received, i.e. the role of `wire_names` in the qubit selection. With this, I mean that we should clarify the following points:
> 
> 1. `wire_names` is absolutely not limiting the qubits used for placing or routing, if these non-trivial passes are included in the transpiler - thus, using `wire_names` as input for a non-trivial placer/routing is only possibly helpful to improve the transpiler solution, but it won't prevent the usage of any qubit in the platform
> 2. `wire_names` are instead exactly honored by the Qibolab backend, when used directly (no transpilation), leading to compilation errors if the connectivity is insufficient
> 3. padding should not be required at any stage, since the information that used to provide is now fully replaced by the usage of `wire_names` and the `connectivity`, to the point that we should even be able to drop the `Preprocessing` (though this should be confirmed)

I have included the updates for 1 and 2.
For 3, each routing/placement pass requires that the length of `wire_names` matches the number of nodes in the connectivity. Is it better to comment that the user must either use `on_qubits` or apply padding before running the transpiler?

File: doc/source/api-reference/qibo.rst [hunk 2]
```diff
@@ -2591,61 +2591,39 @@ size circuits you may benefit from single thread per process, thus set
 Backends
 --------
 
-The main calculation engine is defined in the abstract backend object
-:class:`qibo.backends.abstract.Backend`. This object defines the methods
-required by all Qibo models to perform simulation.
+Backend is the main calculation engine to execute circuits.
+Qibo provides backends for quantum simulation on classical hardware and quantum hardware management and control.
```
Comment (renatomello) [inline doc/source/api-reference/qibo.rst:2595]:
```suggestion
Qibo provides backends for quantum simulation on classical hardware, as well as quantum hardware management and control.
```

File: doc/source/api-reference/qibo.rst [hunk 3]
```diff
@@ -2591,61 +2591,39 @@ size circuits you may benefit from single thread per process, thus set
...
+
+    # Switch to the numpy backend
     qibo.set_backend("numpy")
 
-before creating any circuits or gates. The default backend is the first available
-from ``qibojit``, ``pytorch``, ``tensorflow``, ``numpy``.
 
-Some backends support different platforms. For example, the qibojit backend
+If no backend is specified, the default backend is used.
+The default backend is selected from the available options: ``qibojit``, ``numpy``, ``qiboml``, and ``pytorch``.
```
Comment (renatomello) [inline doc/source/api-reference/qibo.rst:2613]: If this is the exact order of priority, then it should be especified.

Comment (csookim) [reply doc/source/api-reference/qibo.rst]: [58ed798](https://github.com/qiboteam/qibo/pull/1531/commits/58ed798a14253d8d0f431bfe3a21ea640c2f76f7)

File: doc/source/code-examples/advancedexamples.rst [hunk 4]
```diff
@@ -145,6 +145,63 @@ however the user may create the full state as follows:
...
+    circuit.add(gates.M(2, 3))
+
+    # Set the physical qubit names
+    circuit.wire_names = ["C", "A", "B", "D"]
+
+    # This is equivalent to:
+    # H gate on qubit "C"
+    # CNOT gate on qubits "C" and "A"
+    # CZ gate on qubits "C" and "B"
+    # M gate on qubit "B" and "D"
```
Comment (renatomello) [inline doc/source/code-examples/advancedexamples.rst:191]: Explicitly say that this is equal to doing `Circuit(4, wire_names=[...])`

File: doc/source/code-examples/advancedexamples.rst [hunk 5]
```diff
@@ -2222,10 +2257,64 @@ Multiple transpilation steps can be implemented using the :class:`qibo.transpile
...
+The transpiler can be attached to the current backend using the ``set_transpiler`` method.
+Once set, the transpiler will automatically transpile the circuit before each circuit execution.
+
+.. testcode:: python
+
+    import qibo
+    from qibo.transpiler.pipeline import Passes
+    from qibo.transpiler.optimizer import Preprocessing
+    from qibo.transpiler.router import ShortestPaths
+    from qibo.transpiler.unroller import Unroller, NativeGates
```
Comment (renatomello) [inline doc/source/code-examples/advancedexamples.rst:2286]:
```suggestion
    from qibo.transpiler import (
        NativeGates,
        Passes,
        Preprocessing,
        ShortestPaths,
        Unroller,
    )
```

File: doc/source/code-examples/advancedexamples.rst [hunk 6]
```diff
@@ -145,6 +145,63 @@ however the user may create the full state as follows:
...
+The ``Platform`` class in Qibolab represents a QPU device controlled by one or more instruments.
+By specifying the platform name, the user can select the quantum hardware device for circuit execution.
+When executing the circuit, it will be automatically transpiled using the :ref:`Default Transpiler <tutorials_set_transpiler>`.
+
+.. code-block:: python
+
+    import qibo
+
+    # Set the backend and platform
+    qibo.set_backend("qibolab", platform="qw5q_platinum")
```
Comment (renatomello) [inline doc/source/code-examples/advancedexamples.rst:162]:
```suggestion
    from qibo import set_backend

    # Set the backend and platform
    set_backend("qibolab", platform="qw5q_platinum")
```

File: doc/source/code-examples/advancedexamples.rst [hunk 7]
```diff
@@ -2184,36 +2222,33 @@ Multiple transpilation steps can be implemented using the :class:`qibo.transpile
...
         chip = nx.Graph([("q0", "q2"), ("q1", "q2"), ("q2", "q3"), ("q2", "q4")])
         return chip
 
-    # Define the circuit
-    # wire_names must match nodes in the connectivity graph.
-    # The index in wire_names represents the logical qubit number in the circuit.
-    circuit = Circuit(2, wire_names=["q0", "q1"])
+    # Define a custom set of native gates
+    glist = [gates.GPI2, gates.RZ, gates.Z, gates.CZ]
+    natives = NativeGates(0).from_gatelist(glist)
```
Comment (renatomello) [inline doc/source/code-examples/advancedexamples.rst:2232]:
```suggestion
    gate_list = [gates.GPI2, gates.RZ, gates.Z, gates.CZ]
    natives = NativeGates(0).from_gatelist(gate_list)
```

File: doc/source/code-examples/advancedexamples.rst [hunk 8]
```diff
@@ -2222,10 +2257,64 @@ Multiple transpilation steps can be implemented using the :class:`qibo.transpile
...
+
+How to attach a transpiler to a backend?
+----------------------------------------
+
+The transpiler can be attached to the current backend using the ``set_transpiler`` method.
+Once set, the transpiler will automatically transpile the circuit before each circuit execution.
+
+.. testcode:: python
+
+    import qibo
```
Comment (renatomello) [inline doc/source/code-examples/advancedexamples.rst:2282]:
```suggestion
    from qibo import set_transpiler
```

File: doc/source/code-examples/advancedexamples.rst [hunk 9]
```diff
@@ -2222,10 +2257,64 @@ Multiple transpilation steps can be implemented using the :class:`qibo.transpile
...
+
+* For simulator backends, the default transpiler has no passes, so no transpilation is done.
+
+* For hardware backends, the default transpiler includes the :class:`qibo.transpiler.optimizer.Preprocessing`, :class:`qibo.transpiler.router.Sabre`, and :class:`qibo.transpiler.unroller.Unroller` passes, configured with the backend's connectivity and native gates.
+
+Setting an empty transpiler is equivalent to disabling transpilation.
+
+.. testcode:: python
+
+    import qibo
```
Comment (renatomello) [inline doc/source/code-examples/advancedexamples.rst:2314]:
```suggestion
    from qibo import set_transpiler
```

File: doc/source/code-examples/advancedexamples.rst [hunk 10]
```diff
@@ -2222,10 +2257,64 @@ Multiple transpilation steps can be implemented using the :class:`qibo.transpile
...
+    from qibo.transpiler.pipeline import Passes
+
+    qibo.set_transpiler(Passes())
```
Comment (renatomello) [inline doc/source/code-examples/advancedexamples.rst:2317]:
```suggestion
    from qibo import set_transpiler
    from qibo.transpiler.pipeline import Passes

    set_transpiler(Passes())
```

File: src/qibo/transpiler/asserts.py [hunk 11]
```diff
@@ -29,12 +29,12 @@ def assert_transpiling(
...
-        connectivity (networkx.Graph): chip qubits connectivity.
-        final_layout (dict): final physical-logical qubit mapping.
-        native_gates (NativeGates): native gates supported by the hardware.
-        check_circuit_equivalence (Bool): use simulations to check if the transpiled circuit is the same as the original.
+        original_circuit (:class:`qibo.models.circuit.Circuit`): Circuit before transpiling.
+        transpiled_circuit (:class:`qibo.models.circuit.Circuit`): Circuit after transpiling.
+        connectivity (:class:`networkx.Graph`): Hardware connectivity.
+        final_layout (dict): Final logical-physical qubit mapping.
+        native_gates (:class:`qibo.transpiler.unroller.NativeGates`, optional): Native gates supported by the hardware. Defaults to :class:`qibo.transpiler.unroller.NativeGates.default()`.
+        check_circuit_equivalence (bool, optional): Check if the transpiled circuit is equivalent to the original one. Defaults to :math:`True`.
```
Comment (renatomello) [inline src/qibo/transpiler/asserts.py:37]:
```suggestion
        native_gates (:class:`qibo.transpiler.unroller.NativeGates`, optional): Native gates supported by the hardware. 
            Defaults to :class:`qibo.transpiler.unroller.NativeGates.default()`.
        check_circuit_equivalence (bool, optional): Check if the transpiled circuit is equivalent to the original one.
            Defaults to :math:`True`.
```

File: src/qibo/transpiler/asserts.py [hunk 12]
```diff
@@ -50,32 +50,32 @@ def assert_transpiling(
...
-        transpiled_circuit (:class:`qibo.models.circuit.Circuit`): Transpiled circuit.
-        final_map (dict): logical-physical qubit mapping after routing.
-        test_states (list, optional): states on which the test is performed.
+        original_circuit (:class:`qibo.models.circuit.Circuit`): Circuit before transpiling.
+        transpiled_circuit (:class:`qibo.models.circuit.Circuit`): Circuit after transpiling.
+        final_layout (dict): Final logical-physical qubit mapping.
+        test_states (list, optional): List of states to test the equivalence.
             If ``None``, ``ntests`` random states will be tested. Defauts to ``None``.
-        ntests (int, optional): number of random states tested. Defauts to :math:`3`.
+        ntests (int, optional): Number of random states to test the equivalence. Defaults to ``3``.
```
Comment (renatomello) [inline src/qibo/transpiler/asserts.py:72]:
```suggestion
        ntests (int, optional): Number of random states to test the equivalence. Defaults to :math: `3`.
```

File: src/qibo/transpiler/asserts.py [hunk 13]
```diff
@@ -50,32 +50,32 @@ def assert_transpiling(
...
     """
     backend = NumpyBackend()
     if transpiled_circuit.nqubits != original_circuit.nqubits:
         raise_error(
             ValueError,
-            "Transpiled and original circuit do not have the same number of qubits.",
+            f"Transpiled circuit ({transpiled_circuit.nqubits}) and original circuit ({original_circuit.nqubits}) do not have the same number of qubits.",
```
Comment (renatomello) [inline src/qibo/transpiler/asserts.py:78]:
```suggestion
            f"Transpiled circuit ({transpiled_circuit.nqubits}) and original circuit "
            + f"({original_circuit.nqubits}) do not have the same number of qubits.",
```

File: src/qibo/transpiler/optimizer.py [hunk 14]
```diff
@@ -22,16 +22,18 @@ def __call__(self, circuit: Circuit) -> Circuit:
...
 
         physical_qubits = self.connectivity.number_of_nodes()
         logical_qubits = circuit.nqubits
         if logical_qubits > physical_qubits:
             raise_error(
                 ValueError,
-                "The number of qubits in the circuit can't be greater "
-                + "than the number of physical qubits.",
+                # "The number of qubits in the circuit can't be greater "
+                # + "than the number of physical qubits.",
```
Comment (renatomello) [inline src/qibo/transpiler/optimizer.py:34]:
```suggestion
```

File: src/qibo/transpiler/optimizer.py [hunk 15]
```diff
@@ -22,16 +22,18 @@ def __call__(self, circuit: Circuit) -> Circuit:
...
+                f"The number of qubits in the circuit ({logical_qubits}) "
+                f"can't be greater than the number of physical qubits ({physical_qubits}).",
```
Comment (renatomello) [inline src/qibo/transpiler/optimizer.py:36]:
```suggestion
                f"The number of qubits in the circuit ({logical_qubits}) "
                + f"can't be greater than the number of physical qubits ({physical_qubits}).",
```

File: src/qibo/transpiler/pipeline.py [hunk 16]
```diff
@@ -48,14 +48,12 @@ class Passes:
...
-            Defaults to ``None``.
-        connectivity (:class:`networkx.Graph`, optional): physical qubits connectivity.
-        native_gates (:class:`qibo.transpiler.unroller.NativeGates`, optional): native gates.
-            Defaults to :math:`qibo.transpiler.unroller.NativeGates.default`.
-        on_qubits (list, optional): list of physical qubits to be used.
-            If "None" all qubits are used. Defaults to ``None``.
+        passes (list, optional): List of transpiler passes to be applied sequentially.
+            If ``None``, default transpiler will be used. Defaults to ``None``.
+        connectivity (:class:`networkx.Graph`, optional): Hardware connectivity.
+        native_gates (:class:`qibo.transpiler.unroller.NativeGates`, optional): Native gates supported by the hardware. Defaults to :class:`qibo.transpiler.unroller.NativeGates.default()`.
```
Comment (renatomello) [inline src/qibo/transpiler/pipeline.py:54]:
```suggestion
        native_gates (:class:`qibo.transpiler.unroller.NativeGates`, optional): Native gates supported by the hardware. 
            Defaults to :class:`qibo.transpiler.unroller.NativeGates.default()`.
```

File: src/qibo/transpiler/unroller.py [hunk 17]
```diff
@@ -85,16 +85,7 @@ def from_gate(cls, gate):
...
-
-    Args:
-        native_gates (:class:`qibo.transpiler.unroller.NativeGates`): native gates to use
-            in the transpiled circuit.
-        backend (:class:`qibo.backends.Backend`): backend to use for gate matrix.
-
-    Returns:
-        (:class:`qibo.models.circuit.Circuit`): equivalent circuit with native gates.
-    """
+    """Decomposes a circuit to native gates."""
```
Comment (renatomello) [inline src/qibo/transpiler/unroller.py:88]: Where did the docstring go?

Comment (csookim) [reply src/qibo/transpiler/unroller.py:88]: It has been moved to the docstring of `__init__` and `__call__`.

Comment (csookim) [reply src/qibo/transpiler/unroller.py:88]: [8e903f0](https://github.com/qiboteam/qibo/pull/1531/commits/8e903f02983393a18e42d29a32165f053c33c7d9)

File: doc/source/code-examples/advancedexamples.rst (hunk 6 above)
Comment (csookim) [reply doc/source/code-examples/advancedexamples.rst]: [8e903f0](https://github.com/qiboteam/qibo/pull/1531/commits/8e903f02983393a18e42d29a32165f053c33c7d9)

File: doc/source/code-examples/advancedexamples.rst (hunk 4 above)
Comment (csookim) [reply doc/source/code-examples/advancedexamples.rst]: [8e903f0](https://github.com/qiboteam/qibo/pull/1531/commits/8e903f02983393a18e42d29a32165f053c33c7d9)

File: doc/source/code-examples/advancedexamples.rst (hunk 6 above)
Comment (csookim) [reply doc/source/code-examples/advancedexamples.rst]: [c0f5551](https://github.com/qiboteam/qibo/pull/1531/commits/c0f55511f7aad59162c96c20d0365f80a574967e)

File: doc/source/code-examples/advancedexamples.rst [hunk 18]
```diff
@@ -2222,10 +2259,67 @@ Multiple transpilation steps can be implemented using the :class:`qibo.transpile
...
+    custom_pipeline = Passes(custom_passes, connectivity=star_connectivity())
+
+    # Attach the transpiler to the current backend
+    set_transpiler(custom_pipeline)
+
+If the user does not explicitly set a transpiler, the default transpiler is used.
+
+* For simulator backends, the default transpiler has no passes, so no transpilation is done.
+
+* For hardware backends, the default transpiler includes the :class:`qibo.transpiler.optimizer.Preprocessing`, :class:`qibo.transpiler.router.Sabre`, and :class:`qibo.transpiler.unroller.Unroller` passes, configured with the backend's connectivity and native gates.
```
Comment (renatomello) [inline doc/source/code-examples/advancedexamples.rst:2313]:
```suggestion
* For hardware backends, the default transpiler includes the :class:`qibo.transpiler.optimizer.Preprocessing`, 
:class:`qibo.transpiler.router.Sabre`, and :class:`qibo.transpiler.unroller.Unroller` passes, 
configured with the backend's connectivity and native gates.
```

File: doc/source/code-examples/advancedexamples.rst [hunk 19]
```diff
@@ -145,6 +145,65 @@ however the user may create the full state as follows:
...
+The ``Platform`` class in Qibolab represents a QPU device controlled by one or more instruments.
+By specifying the platform name, the user can select the quantum hardware device for circuit execution.
+When executing the circuit, it will be automatically transpiled using the :ref:`Default Transpiler <tutorials_set_transpiler>`.
+
+.. code-block:: python
+
+    import qibo
+
+    # Set the backend and platform
+    qibo.set_backend("qibolab", platform="dummy")
```
Comment (renatomello) [inline doc/source/code-examples/advancedexamples.rst:162]:
```suggestion
    from qibo import set_backend

    # Set the backend and platform
    set_backend("qibolab", platform="dummy")
```

**Review by renatomello — APPROVED** [review]

**Review by alecandido — APPROVED** [review]

**Pull request merged**